File: String/681. Next Closest Time.py
# Simulation
class Solution:
    def nextClosestTime(self, time: str) -> str:
        cur = 60 * int(time[:2]) + int(time[3:])
        allowed = {int(x) for x in time if x != ':'}
        while True:
            cur = (cur + 1) % (24 * 60)
            if all(digit in allowed
                    for block in divmod(cur, 60) #divmod return (quotient, remainder)
                    for digit in divmod(block, 10)):
                return "{:02d}:{:02d}".format(*divmod(cur, 60))



# Try each combination
class Solution:
    def nextClosestTime(self, time: str) -> str:
        ans = start = 60 * int(time[:2]) + int(time[3:])
        elapsed = 24 * 60
        allowed = {int(x) for x in time if x != ':'}
        for h1, h2, m1, m2 in itertools.product(allowed, repeat = 4): 
            #itertools.product(allowed, repeat = 4) let all four number do the combination
            hours, mins = 10 * h1 + h2, 10 * m1 + m2
            if hours < 24 and mins < 60:
                cur = hours * 60 + mins
                cand_elapsed = (cur - start) % (24 * 60)
                if 0 < cand_elapsed < elapsed:
                    ans = cur
                    elapsed = cand_elapsed

        return "{:02d}:{:02d}".format(*divmod(ans, 60))


class Solution:
    def nextClosestTime(self, time: str) -> str:
        digits = set(time)
        while True:
            time = (datetime.strptime(time, '%H:%M') + timedelta(minutes=1)).strftime('%H:%M')
            if set(time) <= digits: # <= only if time is a subset of digits
                return time


class Solution:
    def nextClosestTime(self, time: str) -> str:
        return min((t <= time, t)
               for i in range(24 * 60)
               for t in ['%02d:%02d' % divmod(i, 60)]
               if set(t) <= set(time))[1]


class Solution:
    def nextClosestTime(self, time: str) -> str:
        hour, minute = time.split(":")
        
        # Generate all possible 2 digit values
        # There are at most 16 sorted values here
        nums = sorted(set(hour + minute))
        two_digit_values = [a+b for a in nums for b in nums]
        
        # Check if the next valid minute is within the hour
        i = two_digit_values.index(minute)
        if i + 1 < len(two_digit_values) and two_digit_values[i+1] < "60":
            return hour + ":" + two_digit_values[i+1]

        # Check if the next valid hour is within the day
        i = two_digit_values.index(hour)
        if i + 1 < len(two_digit_values) and two_digit_values[i+1] < "24":
            return two_digit_values[i+1] + ":" + two_digit_values[0]
        
        # Return the earliest time of the next day
        return two_digit_values[0] + ":" + two_digit_values[0]

File: String/681. Next Closest Time.cpp
class Solution {
public:
    string nextClosestTime(string time) {
        set<int>st; 
        for(int i = 0; i<5; ++i){
            if(i == 2) continue;
            st.insert(time[i]-'0');
        }
        string res = time;
        char low = *st.begin() + '0';
        for(int i = 4; i>=0; --i){
            if(i == 2) continue;
            auto it = st.upper_bound(time[i] - '0');
            
            if(it != st.end() && (i == 4 || i == 3 && *it < 6 || i==1 && ( res[0] != '2' || *it < 4 ) || i == 0 && *it < 3)){
                
                res[i] = '0' + *it;
                for(int j = i+1; j<=4; ++j){
                    if(j==2) continue;
                    res[j] = low;
                }
                return res;
            }
        }
        
        return string(1,low) + string(1,low) + ":" + string(1,low) + string(1,low);
    }
};